### ISSUE TITLE
any plans for Julia?

### ISSUE THREAD

**Issue by dylanmmarshall** [googlecolab/colabtools#965]
Sorry if not the place for the question but after all, CoLab was modeled after Jupyter and Jupyter stands for Julia / Python / R... Am wondering if there are plans for Julia use in CoLab or not?

**Comment by ViralBShah** [comment]
I wonder what the result of the triage was. We in the Julia community are happy to collaborate and help make it happen.

**Comment by dyl4nm4rsh4ll** [comment]
Would be nice as it would mean I wouldn't have to leave my favorite tool 🤗

**Comment by Oblynx** [comment]
There are some instructions going around for running Julia on Colab in a rather hacky way, also with GPUs:
- https://dev117uday.medium.com/how-to-run-julia-on-google-colab-dd631e6e7e43
- https://discourse.julialang.org/t/julia-on-google-colab-free-gpu-accelerated-shareable-notebooks/15319

I tried to follow them but was unsuccessful. Maybe things have changed on the Colab side since these hints came out; after all they're pretty hacky. 
In the meantime I'm giving https://nextjournal.com/ a try :smile:

**Comment by oschulz** [comment]
Now that Julia has matured even more, maybe this could be revisited?

There are (hacky) ways of running Julia on Colab, but they are not user friendly. They demonstrate that it shouldn't be a problem technically at all though - Colab basically would just need to provide Julia binaries and preinstall the "IJulia.jl" package.

**Comment by Dsantra92** [comment]
Seems like the hacky methods are no longer working. Running any julia command using `!` results in error. Tried this with multiple Julia versions (1.7+ and 1.8+ s).

```
!julia -e "a=10"
```
Results in
```bash
munmap_chunk(): invalid pointer

signal (6): Aborted
in expression starting at none:0
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x7fde015d926d)
unknown function (ip: 0x7fde015e12fb)
unknown function (ip: 0x7fde015e154b)
close_unit_1 at /workspace/srcdir/gcc-11.1.0/libgfortran/io/unit.c:742
close_units at /workspace/srcdir/gcc-11.1.0/libgfortran/io/unit.c:800
unknown function (ip: 0x7fde0198ef6a)
unknown function (ip: 0x7fde015928a6)
exit at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
main at julia (unknown line)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x400808)
Allocations: 2722 (Pool: 2712; Big: 10); GC: 0
```

**Comment by SoumenMaity922** [comment]
Now I am Julia 1.8.5 in my local machine, but 
I cannot able to use Julia in google Collab . In the runtime type, there is only one option to change the hardware accelerator. So, how can I use  Julia in collab

**Comment by oschulz** [comment]
Now if Julia were just included in Colab by default ... :-)

**Comment by MilesCranmer** [comment]
Confirmed everything is working thanks to @metrizable and the colab's team help in #3385.

For running Julia, check out @ageron's https://github.com/ageron/julia_notebooks for an example of replacing the Python kernel with Julia, so you can run Julia code directly inside the notebook.

**Comment by oschulz** [comment]
The reload and so on isn't really beginner/classroom-friendly, though. Any chance that Julia could be included in Colab?

**Comment by metrizable** [comment]
@dylanmmarshall A Julia runtime is now available by default in Colab. We've gone with LTS 1.10.8 and pre-installed a few packages by default (such as Dataframes, Plots, CSV), but currently not the CUDA libs for Julia. I imagine these could come in the future.

**Comment by bjarthur** [comment]
Plots?  Makie.jl is more popular now.  but thanks for this!!  super exciting.

![Image](https://github.com/user-attachments/assets/080f040b-cd0a-4903-a0d2-8df65413a7f5)

**Comment by MilesCranmer** [comment]
@bjarthur:
<img width="437" alt="Image" src="https://github.com/user-attachments/assets/d147dcd7-f660-42a4-bf3d-03570d68bfa8" />


(Congrats to the colab team, this is really awesome!)

**Comment by MilesCranmer** [comment]
@3f6a you can just do 

```julia
using Pkg
Pkg.add("CairoMakie")
```

in a cell.

**Comment by oschulz** [comment]
> A Julia runtime is now available by default in Colab.

Thanks a lot, that's so awesome!

> pre-installed a few packages by default [...] but currently not the CUDA libs for Julia.

As an alternative you could configure Julia to use CUDA libs installed in the Colab container, by adding CUDA_Runtime_jll` to the packages in the default `Project.toml`, and placing a `LocalPreferences.toml` file next to the default `Project.toml`, containing

```
[CUDA_Runtime_jll]
local = "true"
```

**Comment by Moelf** [comment]
https://github.com/googlecolab/colabtools/issues/965#issuecomment-2702290744

this is still not addressed, Makie.jl is still installed instead of CairoMakie.jl

**Comment by oschulz** [comment]
> this is still not addressed, Makie.jl is still installed instead of CairoMakie.jl

It would be nice to have all of Makie, MakieCore and CairoMakie available to the  user, I think. Installing CairoMakie pulls them in anyhow, of course, put it couldn't hurt to have them in the default Project.toml explicitly.

Having DifferentiationInterface, ADTypes, ChainRules, ChainRulesCore, Zygote, Enzyme and ForwardDiff in the default Project.toml would be helpful too for didactic notebooks on machine learning. They are installed already anyway (except for DifferentiationInterface), but users currently can't load them directly.

**Comment by oschulz** [comment]
Here's a workaround ([example Colab notebook link](https://colab.research.google.com/drive/1351EjrlNOEbBtBsEqlGfXRiNKamEbDCE)) to install and use additional packages in a "zero setup" (well, close to) way:

```julia
macro installpkgs(expr)
    function extract_pkgnames(ex)
        if !(ex isa Expr)
            Symbol[]
        elseif ex.head in (:using, :import)
            [(p.head == :. ? only(p.args) : only(p.args[1].args))::Symbol for p in ex.args]
        elseif ex.head in (:block, :let)
            vcat(extract_pkgnames.(ex.args)...)
        elseif ex.head in (:if, :elseif, :for, :while)
            vcat(extract_pkgnames.(ex.args[2:end])...)
        else
            Symbol[]
        end
    end
    pkgnames = extract_pkgnames(expr)
    :(import Pkg; Pkg.add($(String.(pkgnames)); preserve=Pkg.PRESERVE_TIERED_INSTALLED); $expr)
end
```

It's basically a more powerful version of `AddPackage.@add`, so users can just do

```julia
@installpkgs begin
    using Distributions
    using Makie
    import CairoMakie
end
```

**Comment by 3f6a** [comment]
Can this macro `@installpkgs` be added as a pre-installed package to Colab notebooks? So that people don't have to write the macro definition each time.

**Comment by oschulz** [comment]
> Can this macro @installpkgs be added as a pre-installed package to Colab notebooks? So that people don't have to write the macro definition each time.

I don't think that would be a stable solution, and it would be Colab-only. But maybe we could have a macro like this in Julia? There's a discussion on such features (JuliaLang/Pkg.jl#3027)?

**Comment by oschulz** [comment]
I've registered a little package AutoPkg now that provides `@autopkg` (an improved version of the `@installpkgs` macro above) to make it easy to add and use packages in one go in Colab notebooks ([example link](https://colab.research.google.com/drive/1351EjrlNOEbBtBsEqlGfXRiNKamEbDCE)) and the like.

Just have a cell like

```julia
import Pkg; Pkg.add("AutoPkg"); using AutoPkg
```

at the top of the notebook, then you can just do

```julia
@autopkg begin
    using Distributions
    using Makie
    import CairoMakie
end
```

but also complex stuff like

```julia
@autopkg begin
    using SomePackage, SomeOtherPackage
    import YetAnotherPackage

    if some_condition
        using OneMorePackage: some_feature
    endI've registered a little package AutoPkg now that includes an improved version of the macro above, to make it easy to add and use packages in one go in Colab notebooks ([example link](https://colab.research.google.com/drive/1351EjrlNOEbBtBsEqlGfXRiNKamEbDCE)) and the like.

Just have a cell like

```julia
import Pkg; Pkg.add("AutoPkg"); using AutoPkg
```

at the top of the notebook, then you can just do

```julia
@autopkg begin
    using Distributions
    using Makie
    import CairoMakie
end
```

but also complex stuff like

```julia
@autopkg begin
    using SomePackage, SomeOtherPackage
    import YetAnotherPackage

    if some_condition
        using OneMorePackage: some_feature
    end
end
```

On Google Colab, `@autopkg` installes packages directly (since the notebooks run in isolated containers anyway), otherwise it activates a temporary Julia project env. Also useful to quickly try out packages on the REPL and so on.
end
```

On Google Colab, `@autopkg` installs packages directly (since the notebooks run in isolated containers anyway), otherwise it activates a temporary Julia project env. Also useful to quickly try out packages on the REPL and so on.
